### PR TITLE
Fixes #36495 - Extend Windows templates for Puppet and Ansible

### PR DESCRIPTION
**Pull request by nadjaheitmann**

### CODE REVIEW

File: app/views/unattended/provisioning_templates/script/windows_default_script.erb [hunk 1]
```diff
@@ -21,13 +24,25 @@ set WGET=wget64.exe
...
 <%= @host.diskLayout %>
 
-echo Started downloading main WIM
 
-%WGET% <%= proxy_string %> "<%= medium_uri %>/sources/images.ini" -O X:\images.ini
+<% if iface.subnet && !iface.subnet.dhcp_boot_mode? %>
+echo Setting network config for static interfaces
+    netsh interface ip set address name="Ethernet0" static "<%= iface.ip %>" "<%= iface.subnet.mask %>" "<%= iface.subnet.gateway %>"
+    net stop "DHCP-Client"
+    net stop "DHCP Client"
```
Comment (sbernhard) [inline app/views/unattended/provisioning_templates/script/windows_default_script.erb:34]: why 2 times?

Comment (jtruestedt) [reply app/views/unattended/provisioning_templates/script/windows_default_script.erb:34]: Because the service has different names depending on Windows-Version (i am not sure if it just the major version or also language-dependency).
Same applies for DNS

File: app/views/unattended/provisioning_templates/provision/windows_default_provisioning.erb [hunk 2]
```diff
@@ -12,13 +12,15 @@ description: |
...
   - systemUILanguage: en-US
   - systemTimeZone: GMT Standard Time #see https://msdn.microsoft.com/en-us/library/ms912391(v=winembedded.11).aspx
-  - wimImageName = Windows 8.1 Pro # Image name seems only necessary if the WIM contains more than one image
+  - wimImageName = Windows 10 Pro # Image name seems only necessary if the WIM contains more than one image
 -%>
 <%
   system_locale = host_param('systemLocale') ? host_param('systemLocale') : 'en-US'
   system_ui_lang = host_param('systemUILanguage') ? host_param('systemUILanguage') : 'en-US'
   setup_ui_lang = 'en-US'
+  input_locale = host_param('inputLocale') ? host_param('inputLocale'): 'de-DE'
```
Comment (jtruestedt) [inline app/views/unattended/provisioning_templates/provision/windows_default_provisioning.erb]: Do you really want to set the german keyboard-layout as the default?
I would use the en-US one as default for the foreman-community

File: app/views/unattended/provisioning_templates/finish/windows_default_finish.erb [hunk 3]
```diff
@@ -33,6 +33,7 @@ description: |
   puppet_enabled = !host_param_true?('skip-puppet-setup') && (host_puppet_server.present? || host_param_true?('force-puppet'))
   salt_enabled = host_param('salt_master') ? true : false
   chef_enabled = @host.respond_to?(:chef_proxy) && @host.chef_proxy
+  network_location = host_param('networklocation') ? host_param('networklocation') : 'Private'
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/finish/windows_default_finish.erb]: You can provide a default to `host_param`
```suggestion
  network_location = host_param('networklocation', 'Private')
```

This is repeated a few times. I'm only commenting once, but please fix it in the other files too.

**Review by ekohl — CHANGES_REQUESTED** [review]

File: app/views/unattended/provisioning_templates/finish/windows_default_finish.erb (hunk 3 above)
Comment (nadjaheitmann) [reply app/views/unattended/provisioning_templates/finish/windows_default_finish.erb]: @ekohl Thanks for reviewing! I was just looking for `host_param` in the templates and I did not find any template that makes use of the default. They are using either the notation above or something like 
```suggestion
  network_location = host_param('networklocation') || 'Private'
```
Readability wise I like that the most, although it kind of makes sense to use the default value if exists. Any thoughts?

File: app/views/unattended/provisioning_templates/provision/windows_default_provisioning.erb (hunk 2 above)
Comment (nadjaheitmann) [reply app/views/unattended/provisioning_templates/provision/windows_default_provisioning.erb]: True, thanks!

File: app/views/unattended/provisioning_templates/finish/windows_default_finish.erb (hunk 3 above)
Comment (ekohl) [reply app/views/unattended/provisioning_templates/finish/windows_default_finish.erb]: Internally it already calls `host.host_param(param_name) || default` so your latter suggestion can also work. I'm still leaning to `host_param('networklocation', 'Private')` to match `.fetch()` but we don't implement block support so I can see arguments for both.

Comment (nadjaheitmann) [reply app/views/unattended/provisioning_templates/finish/windows_default_finish.erb]: I have implemented your suggestion. I feel that it looks a bit strange. 

Finally, I think it would be a good idea to unify the notation for all the templates.

Comment (nadjaheitmann) [reply app/views/unattended/provisioning_templates/finish/windows_default_finish.erb]: @ekohl @sbernhard You outvoted me, we'll go with @ekohl 's suggestion :) I hope I did not miss any occurences. Please have a look if anything else needs changing. I suggest to open an issue for unifying the usage of `host_param` and `host_param_true?` within the unattended templates. Any objections?

File: app/views/unattended/provisioning_templates/finish/windows_default_finish.erb [hunk 4]
```diff
@@ -31,8 +31,9 @@ description: |
 <%
   # safemode renderer does not support unary negation
   puppet_enabled = !host_param_true?('skip-puppet-setup') && (host_puppet_server.present? || host_param_true?('force-puppet'))
-  salt_enabled = host_param('salt_master') ? true : false
+  salt_enabled = !host_param('salt_master').nil?
```
Comment (sbernhard) [inline app/views/unattended/provisioning_templates/finish/windows_default_finish.erb]: hmm, " " would be 'not nil', too. Maybe use this: !host_param('salt_master').to_s.empty?

**Comment by nadjaheitmann** [comment]
@ekohl Anything else that should be improved?

**Comment by nadjaheitmann** [comment]
@ekohl @sbernhard Can we move forward here?

File: app/views/unattended/provisioning_templates/finish/windows_default_finish.erb (hunk 4 above)
Comment (ekohl) [inline app/views/unattended/provisioning_templates/finish/windows_default_finish.erb]: I think you should treat empty strings as not enabled either
```suggestion
  salt_enabled = host_param('salt_master').present?
```

File: app/views/unattended/provisioning_templates/finish/windows_default_finish.erb [hunk 5]
```diff
@@ -44,89 +45,104 @@ description: |
...
+    cmd /C "netsh winhttp set proxy <%= host_param('http-proxy') %>:<%= host_param('http-proxy-port') %>"
+  <% end %>
+
+  <% unless host_param('computerDomain') %>
+    powershell /c "Get-NetConnectionProfile -InterfaceAlias \"Ethernet0\" | Set-NetConnectionProfile -NetworkCategory <%= network_location %>"
   <% end %>
 
-  <% if host_param('ansible_winrm_transport') == 'basic' %>
+  <% if host_param('ansible_user').present? %>
+    <% if host_param('create_ansible_user') %>
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/finish/windows_default_finish.erb]: If you use `host_param_true?` you can set it to `true` on the host group level and `false` on the host level.
```suggestion
    <% if host_param_true?('create_ansible_user') %>
```

File: app/views/unattended/provisioning_templates/finish/windows_default_finish.erb [hunk 6]
```diff
@@ -44,89 +45,104 @@ description: |
...
-  <% end -%>
-
   <% if puppet_enabled %>
-    echo Installing puppet
+    echo Downloading Puppet installer
+    wget "<%= host_param('win_puppet_source') %>" -O C:\puppet-agent-x64-latest.msi
+    echo Configuring Puppet
+    mkdir "C:\ProgramData\PuppetLabs\puppet\etc"
+    echo [main] > "C:\ProgramData\PuppetLabs\puppet\etc\puppet.conf"
+    echo server=http://<%= @host.puppetmaster %> >> "C:\ProgramData\PuppetLabs\puppet\etc\puppet.conf"
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/finish/windows_default_finish.erb]: I think this shouldn't be needed (see below), but if it is: server is the hostname, there's no protocol. Also, `puppet_puppet_server` is the current macro. `@host.puppetmaster is deprecated.
```suggestion
    echo server=<%= host_puppet_server %> >> "C:\ProgramData\PuppetLabs\puppet\etc\puppet.conf"
```

File: app/views/unattended/provisioning_templates/finish/windows_default_finish.erb [hunk 7]
```diff
@@ -44,89 +45,104 @@ description: |
...
-    IF EXIST c:\deploy\user_data.ps1 powershell.exe -OutputFormat text -command c:\deploy\user_data.ps1
-  <% end -%>
-
   <% if puppet_enabled %>
-    echo Installing puppet
+    echo Downloading Puppet installer
+    wget "<%= host_param('win_puppet_source') %>" -O C:\puppet-agent-x64-latest.msi
+    echo Configuring Puppet
+    mkdir "C:\ProgramData\PuppetLabs\puppet\etc"
+    echo [main] > "C:\ProgramData\PuppetLabs\puppet\etc\puppet.conf"
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/finish/windows_default_finish.erb]: Why is this needed now? I suspected the `puppet.msi` (which is still there) would be sufficient.

File: app/views/unattended/provisioning_templates/finish/windows_default_finish.erb [hunk 8]
```diff
@@ -44,89 +45,104 @@ description: |
...
+    echo server=http://<%= @host.puppetmaster %> >> "C:\ProgramData\PuppetLabs\puppet\etc\puppet.conf"
+    echo autoflush=true >> "C:\ProgramData\PuppetLabs\puppet\etc\puppet.conf"
+    echo Installing Puppet
     start /w "" msiexec /qn /i C:\extras\puppet.msi PUPPET_AGENT_STARTUP_MODE=Manual PUPPET_MASTER_SERVER=<%= host_puppet_server -%> PUPPET_AGENT_ACCOUNT_DOMAIN=<%= @host.domain -%> PUPPET_AGENT_ACCOUNT_USER=administrator PUPPET_AGENT_ACCOUNT_PASSWORD="<%= host_param('domainAdminAccountPasswd') -%>"
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/finish/windows_default_finish.erb]: I suspect `PUPPET_MASTER_SERVER` is wrong here. Reading https://www.puppet.com/docs/puppet/7/install_agents#msi_properties I think it should be:
```suggestion
    start /w "" msiexec /qn /i C:\extras\puppet.msi PUPPET_AGENT_STARTUP_MODE=Manual PUPPET_SERVER=<%= host_puppet_server -%> PUPPET_AGENT_ACCOUNT_DOMAIN=<%= @host.domain -%> PUPPET_AGENT_ACCOUNT_USER=administrator PUPPET_AGENT_ACCOUNT_PASSWORD="<%= host_param('domainAdminAccountPasswd') -%>"
```

It should also set `PUPPET_CA_SERVER=<%= host_puppet_ca_server>%`.

File: app/views/unattended/provisioning_templates/finish/windows_default_finish.erb [hunk 9]
```diff
@@ -44,89 +45,104 @@ description: |
...
-
   <% if puppet_enabled %>
-    echo Installing puppet
+    echo Downloading Puppet installer
+    wget "<%= host_param('win_puppet_source') %>" -O C:\puppet-agent-x64-latest.msi
+    echo Configuring Puppet
+    mkdir "C:\ProgramData\PuppetLabs\puppet\etc"
+    echo [main] > "C:\ProgramData\PuppetLabs\puppet\etc\puppet.conf"
+    echo server=http://<%= @host.puppetmaster %> >> "C:\ProgramData\PuppetLabs\puppet\etc\puppet.conf"
+    echo autoflush=true >> "C:\ProgramData\PuppetLabs\puppet\etc\puppet.conf"
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/finish/windows_default_finish.erb]: I had to look this up, but https://www.puppet.com/docs/puppet/7/configuration.html#autoflush says this should be default:
```suggestion
```

File: app/views/unattended/provisioning_templates/finish/windows_default_finish.erb [hunk 10]
```diff
@@ -44,89 +45,104 @@ description: |
...
-
-  <% if foreman_url('user_data') %>
-    echo execute user data script
-    IF EXIST c:\deploy\user_data.ps1 powershell.exe -OutputFormat text -command c:\deploy\user_data.ps1
-  <% end -%>
-
   <% if puppet_enabled %>
-    echo Installing puppet
+    echo Downloading Puppet installer
+    wget "<%= host_param('win_puppet_source') %>" -O C:\puppet-agent-x64-latest.msi
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/finish/windows_default_finish.erb]: This doesn't match the path below (`C:\extras\puppet.msi`) so likely to cause problems.

File: app/views/unattended/provisioning_templates/finish/windows_default_finish.erb [hunk 11]
```diff
@@ -44,89 +45,104 @@ description: |
...
+    echo Installing Puppet
     start /w "" msiexec /qn /i C:\extras\puppet.msi PUPPET_AGENT_STARTUP_MODE=Manual PUPPET_MASTER_SERVER=<%= host_puppet_server -%> PUPPET_AGENT_ACCOUNT_DOMAIN=<%= @host.domain -%> PUPPET_AGENT_ACCOUNT_USER=administrator PUPPET_AGENT_ACCOUNT_PASSWORD="<%= host_param('domainAdminAccountPasswd') -%>"
-    echo set puppet to auto start
+    echo Setting Puppet to auto start
     sc config puppet start= auto
     sc query puppet
   <% end%>
 
-  <% if host_param('foremanDebug') != true %>
+  <% if !host_param_true('foremanDebug') %>
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/finish/windows_default_finish.erb]:
```suggestion
  <% if !host_param_true?('foremanDebug') %>
```

File: app/views/unattended/provisioning_templates/finish/windows_default_finish.erb [hunk 12]
```diff
@@ -44,89 +45,104 @@ description: |
...
 
-    echo Safely remove leftover directories
-    sdelete.exe -accepteula -p 2 -r c:\drivers
-    sdelete.exe -accepteula -p 2 -r c:\updates
+    echo Removing leftover directories
+    rd /s /q c:\MININT
+    rd /s /q c:\drivers
+    rd /s /q c:\updates
+
+    <% if puppet_enabled %>
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/finish/windows_default_finish.erb]:
```suggestion
    <% if puppet_enabled -%>
```

**Review by ekohl — CHANGES_REQUESTED** [review]

File: app/views/unattended/provisioning_templates/finish/windows_default_finish.erb [hunk 13]
```diff
@@ -44,89 +45,104 @@ description: |
...
+        echo Performing unsecure domain join
         powershell.exe -OutputFormat text -command Add-Computer -Domain '<%= host_param('computerDomain') -%>' -Options UnsecuredJoin,PasswordPass -Credential (New-Object -TypeName System.Management.Automation.PSCredential $null, (ConvertTo-SecureString -String '<%= host_param('machinePassword') -%>' -AsPlainText -Force))
       <% end %>
     <% end %>
   <% end %>
 
+  <%#
+  Adjust power setting to 'High performance'
+  -%>
+  powershell /c "powercfg /setactive 8c5e7fda-e8bf-4a96-9a85-a6e23a8c635c"
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/finish/windows_default_finish.erb]: Is this really appropriate for this PR? I can't speak for Windows admins since I rarely touch it, but doing this unconditionally may not be correct for users.

**Review by ekohl — DISMISSED** [review]
Please consistently use `-%>`, at least for newly added blocks.

**Comment by nadjaheitmann** [comment]
> Please consistently use `-%>`, at least for newly added blocks.

Thanks, I'll try to unify the templates. Working on the other changes as well, thank you for your input! Highly appreciated!

**Comment by nadjaheitmann** [comment]
@ekohl I cleaned up the tags (hope I got them all) and modified the configurations in question. Right now, there are three commits for better visibility of the changes. Do you mind having another look?

**Comment by nadjaheitmann** [comment]
@ekohl Anything else that needs fixing?

File: test/unit/foreman/renderer/snapshots/ProvisioningTemplate/finish/Windows_default_finish.windows10_dhcp.snap.txt [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+    rd /s /q c:\drivers
+    rd /s /q c:\updates
+
+            echo Removing Puppet installer
+        sdelete.exe -accepteula -p 2 C:\puppet-agent-x64-latest.msi
+    
+    echo Removing deploy directory
+    rd /s /q c:\deploy
+
+
```
Comment (nadjaheitmann) [inline test/unit/foreman/renderer/snapshots/ProvisioningTemplate/finish/Windows_default_finish.windows10_dhcp.snap.txt]: Indentation-wise, this template snapshot looks kind of funny....

Comment (nadjaheitmann) [reply test/unit/foreman/renderer/snapshots/ProvisioningTemplate/finish/Windows_default_finish.windows10_dhcp.snap.txt]: fixed indentation

File: app/views/unattended/provisioning_templates/finish/windows_default_finish.erb (hunk 8 above)
Comment (sbernhard) [reply app/views/unattended/provisioning_templates/finish/windows_default_finish.erb]: Can you add this PUPPET_CA_SERVER @nadjaheitmann ? Does this make sense?

**Review by sbernhard — DISMISSED** [review]

File: app/views/unattended/provisioning_templates/finish/windows_default_finish.erb [hunk 15]
```diff
@@ -31,102 +33,108 @@ description: |
...
+<%= snippet 'windows_network' %>
+
+<% if host_param('ntpServer') -%>
+echo Setting time server
+w32tm /config /manualpeerlist:<%= host_param('ntpServer') %> /syncfromflags:manual /update
+<% end -%>
+
+echo Syncing time
+w32tm /resync
+w32tm /resync
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/finish/windows_default_finish.erb]: Is it intentional that you call this twice?

File: app/views/unattended/provisioning_templates/finish/windows_default_finish.erb [hunk 16]
```diff
@@ -31,102 +33,108 @@ description: |
...
+
+<% if host_param('computerDomain') -%>
+<% if host_param('domainAdminAccount').present? && host_param('domainAdminAccountPasswd').present? -%>
+echo Performing secure domain join
+powershell.exe -OutputFormat text -command Add-Computer -DomainName '<%= host_param('computerDomain') -%>' -Credential (New-Object -TypeName System.Management.Automation.PSCredential '<%= host_param('domainAdminAccount') -%>', (ConvertTo-SecureString -String '<%= host_param('domainAdminAccountPasswd') -%>' -AsPlainText -Force)) <% if host_param('computerOU').present? -%>-OUPath '<%= host_param('computerOU') -%>'<% end -%>
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/finish/windows_default_finish.erb]: Nit: perhaps avoid trailing whitespace?
```suggestion
powershell.exe -OutputFormat text -command Add-Computer -DomainName '<%= host_param('computerDomain') -%>' -Credential (New-Object -TypeName System.Management.Automation.PSCredential '<%= host_param('domainAdminAccount') -%>', (ConvertTo-SecureString -String '<%= host_param('domainAdminAccountPasswd') -%>' -AsPlainText -Force))<% if host_param('computerOU').present? -%> -OUPath '<%= host_param('computerOU') -%>'<% end -%>
```

File: app/views/unattended/provisioning_templates/finish/windows_default_finish.erb [hunk 17]
```diff
@@ -31,102 +33,108 @@ description: |
...
+<% if host_param('remote_desktop') -%>
+cmd /c "netsh advfirewall firewall set rule group=\"remote desktop\" new enable=Yes"
+cmd /c "netsh advfirewall firewall set rule group=\"remotedesktop\" new enable=Yes"
+<% end -%>
+
+<% if puppet_enabled -%>
+echo Downloading Puppet installer
+wget "<%= host_param('win_puppet_source') %>" -O C:\puppet-agent-x64-latest.msi
+echo Installing Puppet
+start /w "" msiexec /qn /i C:\puppet-agent-x64-latest.msi PUPPET_AGENT_STARTUP_MODE=Manual PUPPET_SERVER=<%= host_puppet_server -%> PUPPET_CA_SERVER=<%= host_puppet_ca_server -%> PUPPET_AGENT_ACCOUNT_DOMAIN=<%= @host.domain -%> PUPPET_AGENT_ACCOUNT_USER=administrator PUPPET_AGENT_ACCOUNT_PASSWORD="<%= host_param('domainAdminAccountPasswd') -%>"
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/finish/windows_default_finish.erb]: With the eye on non-x64 support (like ARM), would this be better?
```suggestion
wget "<%= host_param('win_puppet_source') %>" -O C:\puppet-agent-latest.msi
echo Installing Puppet
start /w "" msiexec /qn /i C:\puppet-agent-latest.msi PUPPET_AGENT_STARTUP_MODE=Manual PUPPET_SERVER=<%= host_puppet_server -%> PUPPET_CA_SERVER=<%= host_puppet_ca_server -%> PUPPET_AGENT_ACCOUNT_DOMAIN=<%= @host.domain -%> PUPPET_AGENT_ACCOUNT_USER=administrator PUPPET_AGENT_ACCOUNT_PASSWORD="<%= host_param('domainAdminAccountPasswd') -%>"
```

File: app/views/unattended/provisioning_templates/finish/windows_default_finish.erb [hunk 18]
```diff
@@ -31,102 +33,108 @@ description: |
...
+sdelete.exe -accepteula -p 2 C:\Windows\Setup\Scripts\SetupComplete.cmd
+
+echo Removing leftover directories
+rd /s /q c:\MININT
+rd /s /q c:\drivers
+rd /s /q c:\updates
+
+<% if puppet_enabled -%>
+echo Removing Puppet installer
+sdelete.exe -accepteula -p 2 C:\puppet-agent-x64-latest.msi
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/finish/windows_default_finish.erb]: And if you change the filename above, then here too
```suggestion
sdelete.exe -accepteula -p 2 C:\puppet-agent-latest.msi
```

File: app/views/unattended/provisioning_templates/provision/windows_default_provisioning.erb [hunk 19]
```diff
@@ -70,9 +74,23 @@ description: |
         <component name="Microsoft-Windows-Shell-Setup" processorArchitecture="amd64" publicKeyToken="31bf3856ad364e35" language="neutral" versionScope="nonSxS" xmlns:wcm="http://schemas.microsoft.com/WMIConfig/2002/State" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
             <UserAccounts>
                 <AdministratorPassword>
-                    <PlainText>false</PlainText>
                     <Value><%= root_pass %></Value>
+                    <PlainText>false</PlainText>
                 </AdministratorPassword>
+            <% if host_param('ansible_user') and host_param('create_ansible_user') -%>
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/provision/windows_default_provisioning.erb]: In Ruby `&&` is preferred over `and`
```suggestion
            <% if host_param('ansible_user') && host_param('create_ansible_user') -%>
```

File: app/views/unattended/provisioning_templates/provision/windows_default_provisioning.erb [hunk 20]
```diff
@@ -70,9 +74,23 @@ description: |
...
+                <LocalAccounts>
+                    <LocalAccount wcm:action="add">
+                        <Password>
+                            <Value><%= host_param('ansible_ssh_pass') %></Value>
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/provision/windows_default_provisioning.erb]: I wonder if anyone has `<` in their password and you need to XML escape values

File: app/views/unattended/provisioning_templates/script/windows_default_script.erb [hunk 21]
```diff
@@ -4,15 +4,20 @@ name: Windows peSetup.cmd
...
   See https://community.theforeman.org/t/windows-provisioning-made-easy/16756/
   params:
   - wimImageName: Windows 8.1 Pro # name of wim image to apply
+  - windowsLicenseKey: ABCDE-ABCDE-ABCDE-ABCDE-ABCDE # Valid Windows license key
+  - windowsLicenseOwner: Company, INC # Legal owner of the Windows license key
 -%>
 <%
-  proxy_uri = host_param('http-proxy') ? "http://#{host_param('http-proxy')}:#{host_param('http-proxy-port') || 3128}" : nil
+  iface = @host.provision_interface
+  proxy_uri = host_param('http-proxy') ? "#{host_param('http-proxy')}:#{host_param('http-proxy-port')}" : nil
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/script/windows_default_script.erb]: Are you intentionally dropping the default value?
```suggestion
  proxy_uri = host_param('http-proxy') ? "#{host_param('http-proxy')}:#{host_param('http-proxy-port', 3128)}" : nil
```

File: test/unit/foreman/renderer/snapshots/ProvisioningTemplate/script/Windows_peSetup.cmd.windows10_dhcp.snap.txt [hunk 22]
```diff
@@ -0,0 +1,97 @@
+
+@setlocal enableextensions enabledelayedexpansion
+@echo off
+set WGET=wget64.exe
+
+zerombr
```
Comment (ekohl) [inline test/unit/foreman/renderer/snapshots/ProvisioningTemplate/script/Windows_peSetup.cmd.windows10_dhcp.snap.txt]: Looks like it's using a Red Hat partitioning table. That's probably a bug in the factory.

File: app/services/foreman/template_snapshot_service.rb [hunk 23]
```diff
@@ -205,6 +209,14 @@ def rocky9_dhcp
       define_host_params(host)
     end
 
+    def windows10_dhcp
+      host = FactoryBot.build(:host_for_snapshots_ipv4_dhcp_windows10,
```
Comment (ekohl) [inline app/services/foreman/template_snapshot_service.rb:213]: As noted earlier: this ends up with a Red Hat partition table. Perhaps this is also a problem for the Debian factories.

**Review by ekohl — COMMENTED** [review]
Overall I think this looks pretty good and many notes are more nit picks.

File: app/views/unattended/provisioning_templates/finish/windows_default_finish.erb (hunk 18 above)
Comment (nadjaheitmann) [reply app/views/unattended/provisioning_templates/finish/windows_default_finish.erb]: Note: If I change the name here, I also need to change it in the windows user data template.

File: test/unit/foreman/renderer/snapshots/ProvisioningTemplate/finish/Windows_default_finish.windows10_dhcp.snap.txt [hunk 24]
```diff
@@ -0,0 +1,78 @@
...
+powershell /c "Get-NetConnectionProfile -InterfaceAlias \"Ethernet0\" | Set-NetConnectionProfile -NetworkCategory Private"
+
+
+
+
+echo Downloading Puppet installer
+wget "" -O C:\puppet-agent-x64-latest.msi
+echo Installing Puppet
+start /w "" msiexec /qn /i C:\puppet-agent-x64-latest.msi PUPPET_AGENT_STARTUP_MODE=Manual PUPPET_SERVER= PUPPET_CA_SERVER= PUPPET_AGENT_ACCOUNT_DOMAIN=snap.example.com PUPPET_AGENT_ACCOUNT_USER=administrator PUPPET_AGENT_ACCOUNT_PASSWORD=""
```
Comment (nadjaheitmann) [inline test/unit/foreman/renderer/snapshots/ProvisioningTemplate/finish/Windows_default_finish.windows10_dhcp.snap.txt]: What looks funny to me is that `PUPPET_SERVER=` and `PUPPET_CA_SERVER=` are set without any values here. @ekohl Is this okay for the test or is it possible to set default values for host_puppet_server and host_puppet_ca_server?

File: app/services/foreman/template_snapshot_service.rb (hunk 23 above)
Comment (nadjaheitmann) [reply app/services/foreman/template_snapshot_service.rb:213]: I found that there is a partition table configured for Debian systems but I have not checked how thoroughly it is used. https://github.com/theforeman/foreman/blob/develop/test/factories/host_related.rb#L23

File: test/unit/foreman/renderer/snapshots/ProvisioningTemplate/script/Windows_peSetup.cmd.windows10_dhcp.snap.txt (hunk 22 above)
Comment (nadjaheitmann) [reply test/unit/foreman/renderer/snapshots/ProvisioningTemplate/script/Windows_peSetup.cmd.windows10_dhcp.snap.txt]: I have added a partition table for windows. I am absolutely not sure if I added a correct format, though.

File: app/views/unattended/provisioning_templates/finish/windows_default_finish.erb (hunk 15 above)
Comment (nadjaheitmann) [reply app/views/unattended/provisioning_templates/finish/windows_default_finish.erb]: I removed the second call.

File: app/views/unattended/provisioning_templates/provision/windows_default_provisioning.erb (hunk 20 above)
Comment (nadjaheitmann) [reply app/views/unattended/provisioning_templates/provision/windows_default_provisioning.erb]: Yeah, you are right, we need to fix this as well.

Comment (nadjaheitmann) [reply app/views/unattended/provisioning_templates/provision/windows_default_provisioning.erb]: Added HTML escaping for Windows https://www.w3.org/TR/REC-xml/#sec-cdata-sect

**Comment by nadjaheitmann** [comment]
> Overall I think this looks pretty good and many notes are more nit picks.

Thanks again @ekohl for the review. I addressed all the issues you mentioned. Do you mind having another look?

**Comment by nadjaheitmann** [comment]
@ekohl More changes required? :)

**Review by stejskalleos — APPROVED** [review]

**Comment by stejskalleos** [comment]
Thanks @nadjaheitmann !
